Zend/zend_opcode.c

```diff
@@ -236,6 +236,15 @@ ZEND_API void destroy_zend_class(zval *zv)
 				zval *end = p + ce->default_static_members_count;
 
 				while (p != end) {
+					if (UNEXPECTED(Z_ISREF_P(p))) {
+						zend_property_info *prop_info;
+						ZEND_REF_FOREACH_TYPE_SOURCES(Z_REF_P(p), prop_info) {
+							if (prop_info->ce == ce && p - ce->default_static_members_table == prop_info->offset) {
```

Comment:
Same issue with CE checks.

Reply:
Is there really an issue with statics? If I read the code correctly, the static variable is always stored alongside the actual defining class and you are not allowed to redeclare statics either (visibility change not possible), thus this should be fine.

Reply:
I think you are basically right, but there might still be an issue if a static property is inherited from an internal class into a user class, in which case we will duplicate the property info.

Reply:
This is correct, though we don't have internal typed static properties yet…
I think the solution would be to rather _not_ duplicate static property type info. I see no reason why we are even doing that.

Reply:
@nikic Please check out https://github.com/php/php-src/compare/master...bwoebi:prevent_property_info_duplication - this should keep this simple and it will also make the other check much simpler as we can simply assume there will be at most two different prop_infos pointing to the same property offset.
It would simplify the whole loop to:

```C
zend_property_info *prop_info, *target_info;
ZEND_REF_FOREACH_TYPE_SOURCES(Z_REF_P(p), prop_info) {
	if (p == OBJ_PROP(object, prop_info->offset) && instanceof_function(object->ce, prop_info->ce)) {
		if (prop_info->flags & (ZEND_ACC_PUBLIC | ZEND_ACC_PRIVATE)) {
			goto found;
		} else {
			target_info = prop_info;
		}
	}
} ZEND_REF_FOREACH_TYPE_SOURCES_END();
prop_info = target_info;
found:
ZEND_REF_DEL_TYPE_SOURCE(Z_REF_P(p), prop_info);
```

Reply:
@bwoebi I don't think I understand the purpose of that change. Is it only about the static property case?

At least right now (and that patch does not seem to change this), if you override a public property with another public property that will also result in a new property_info, it's not necessarily due to a visibility change.

Reply:
Eih, I forgot about that. And we cannot just overwrite it because, well, the attached doc comment...
But still, it's applicable to the static property case.